### PR TITLE
feat(node): implement fs.statfs()

### LINKED ISSUE #21139: Missing statfs in `node:fs`
See https://github.com/denoland/wanted_modules/issues/85

### PR DESCRIPTION
**Pull request by nayeemrmn**
Closes #21139.

### CODE REVIEW

File: ext/node/ops/fs.rs [hunk 1]
```diff
@@ -78,3 +79,156 @@ where
...
+/// precision may be lost if they are parsed to numbers by serde_v8.
+pub struct StatFs {
+  #[serde(rename = "type")]
+  pub typ: String,
+  pub bsize: String,
+  pub blocks: String,
+  pub bfree: String,
+  pub bavail: String,
+  pub files: String,
+  pub ffree: String,
```
Comment (littledivy) [inline ext/node/ops/fs.rs:94]: Use `u64`. serde_v8 will serialize it has BigInt if it doesn't fit. Source: https://docs.rs/serde_v8/latest/src/serde_v8/ser.rs.html#440-449

File: ext/node/polyfills/_fs/_fs_statfs.ts [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+    this.blocks = blocks;
+    this.bfree = bfree;
+    this.bavail = bavail;
+    this.files = files;
+    this.ffree = ffree;
+  }
+}
+
+export function statfs(path, options, callback) {
+  if (typeof options === "function") {
```
Comment (littledivy) [inline ext/node/polyfills/_fs/_fs_statfs.ts:27]: This file is not typed. Please rename to `.js` or add typings.

File: ext/node/ops/fs.rs [hunk 3]
```diff
@@ -78,3 +79,156 @@ where
...
+  {
+    use std::ffi::OsStr;
+    use std::os::unix::ffi::OsStrExt;
+
+    let path = OsStr::new(&path);
+    let mut cpath = path.as_bytes().to_vec();
+    cpath.push(0);
+    if bigint {
+      #[cfg(not(target_os = "macos"))]
+      // SAFETY: Normal statfs usage.
```
Comment (littledivy) [inline ext/node/ops/fs.rs:126]:
```suggestion
      // SAFETY: `cpath` is NUL-terminated and result is pointer to valid statfs memory.
```

Nit

File: ext/node/ops/fs.rs [hunk 4]
```diff
@@ -78,3 +79,156 @@ where
...
+      let (code, result) = unsafe {
+        let mut result: libc::statfs64 = std::mem::zeroed();
+        (libc::statfs64(cpath.as_ptr() as _, &mut result), result)
+      };
+      #[cfg(target_os = "macos")]
+      // SAFETY: Normal statfs usage.
```
Comment (littledivy) [inline ext/node/ops/fs.rs:132]: Ditto

File: ext/node/ops/fs.rs [hunk 5]
```diff
@@ -78,3 +79,156 @@ where
...
+        (libc::statfs(cpath.as_ptr() as _, &mut result), result)
+      };
+      if code == -1 {
+        return Err(
+          std::io::Error::new(
+            std::io::ErrorKind::NotFound,
+            "No such file or directory",
+          )
+          .into(),
+        );
```
Comment (littledivy) [inline ext/node/ops/fs.rs:144]: Use `std::io::Error::last_os_error()`

File: ext/node/polyfills/internal/primordials.mjs [hunk 6]
```diff
@@ -12,6 +12,7 @@ export const ArrayPrototypeSlice = (that, ...args) => that.slice(...args);
 export const ArrayPrototypeSome = (that, ...args) => that.some(...args);
 export const ArrayPrototypeSort = (that, ...args) => that.sort(...args);
 export const ArrayPrototypeUnshift = (that, ...args) => that.unshift(...args);
+export const BigInt = globalThis.BigInt;
```
Comment (bartlomieju) [inline ext/node/polyfills/internal/primordials.mjs:15]: Is `BigInt` not available in `primordials` from `ext:core/mod.js`? If so we should fix it 😬

**Review by bartlomieju — APPROVED** [review]
LGTM, @littledivy any objections?

File: ext/node/ops/fs.rs [hunk 7]
```diff
@@ -78,3 +79,140 @@ where
...
+    use deno_core::anyhow::anyhow;
+    use std::ffi::OsStr;
+    use std::os::windows::ffi::OsStrExt;
+    use windows_sys::Win32::Storage::FileSystem::GetDiskFreeSpaceW;
+
+    let _ = bigint;
+    // Using a vfs here doesn't make sense, it won't align with the windows API
+    // call below.
+    #[allow(clippy::disallowed_methods)]
+    let path = Path::new("Cargo.toml").canonicalize().unwrap();
```
Comment (littledivy) [inline ext/node/ops/fs.rs:178]: Wouldn't `canonicalize()` fail if `Cargo.toml` is not present?

**Review by littledivy — CHANGES_REQUESTED** [review]
https://github.com/denoland/deno/pull/22862#discussion_r1522415190

File: ext/node/ops/fs.rs (hunk 7 above)
Comment (littledivy) [reply ext/node/ops/fs.rs:178]: I'm not sure why we need this. The user `path` argument is shadowed by this hardcoded path to Cargo.toml

Comment (nayeemrmn) [reply ext/node/ops/fs.rs]: I must have copy-pasted this from my windows playground repo... that's frustrating there was no lint error.

**Review by littledivy — APPROVED** [review]
LGTM!

**Pull request merged**